gluten-core/src/main/scala/io/glutenproject/execution/GenerateExecTransformer.scala

```diff
@@ -145,33 +166,57 @@ case class GenerateExecTransformer(
       }
     }
 
-    val relNode = if (childCtx != null) {
-      getRelNode(
-        context,
-        operatorId,
-        child.output,
-        childCtx.root,
-        generatorNode,
-        childOutputNodes,
-        validation = false)
+    val inputRel = if (childCtx != null) {
+      childCtx.root
     } else {
       val attrList = new java.util.ArrayList[Attribute]()
       for (attr <- child.output) {
         attrList.add(attr)
       }
       val readRel = RelBuilder.makeReadRel(attrList, context, operatorId)
-      getRelNode(
-        context,
-        operatorId,
-        child.output,
-        readRel,
-        generatorNode,
-        childOutputNodes,
-        validation = false)
+      readRel
     }
+
+    val projRel = if (BackendsApiManager.isVeloxBackend && needsProjection(generator)) {
```

Comment:
CC @lgbo-ustc I'm adding support for velox and made a small refactor on the input of generate. CK CI shows green so I assume this is safe for CK backend

Reply:
seems not affect CK backend.

Reply:
I think the pre-projection may also could be applied to CK.